services/venice-router/src/main/java/com/linkedin/venice/router/MetaDataHandler.java

```diff
@@ -372,54 +374,45 @@ private void handleUpdateSchemaLookup(ChannelHandlerContext ctx, VenicePathParse
     }
   }
 
-  private void handleD2ServiceLookup(ChannelHandlerContext ctx, VenicePathParserHelper helper, HttpHeaders headers)
+  private void handleD2ServiceLookup(ChannelHandlerContext ctx, VenicePathParserHelper helper, HttpRequest request)
       throws IOException {
     String storeName = helper.getResourceName();
-    checkResourceName(storeName, "/" + TYPE_CLUSTER_DISCOVERY + "/${storeName}");
+    if (StringUtils.isEmpty(storeName)) {
+      Map<String, String> queryParams = helper.extractQueryParameters(request);
+      storeName = queryParams.get(NAME);
+    }
+    checkResourceName(
+        storeName,
+        "/" + TYPE_CLUSTER_DISCOVERY + "/${storeName} or /" + TYPE_CLUSTER_DISCOVERY + "?store_name=${storeName}");
     Optional<StoreConfig> config = storeConfigRepo.getStoreConfig(storeName);
     if (!config.isPresent() || StringUtils.isEmpty(config.get().getCluster())) {
       String errorMsg = "Cluster for store: " + storeName + " doesn't exist";
-      setupErrorD2DiscoveryResponseAndFlush(NOT_FOUND, errorMsg, headers, ctx);
+      setupErrorD2DiscoveryResponseAndFlush(NOT_FOUND, errorMsg, ctx);
       return;
     }
     String clusterName = config.get().getCluster();
     String d2Service = getD2ServiceByClusterName(clusterName);
     if (StringUtils.isEmpty(d2Service)) {
       String errorMsg = "D2 service for store: " + storeName + " doesn't exist";
-      setupErrorD2DiscoveryResponseAndFlush(NOT_FOUND, errorMsg, headers, ctx);
+      setupErrorD2DiscoveryResponseAndFlush(NOT_FOUND, errorMsg, ctx);
       return;
     }
-    String serverD2Service = getServerD2ServiceByClusterName(clusterName);
-    if (headers.contains(D2_SERVICE_DISCOVERY_RESPONSE_V2_ENABLED)) {
```

Comment:
Hmm, I got a question to confirm for this V2->V1 migration: If an existing client is sending d2 discovery request via v2 and received by new Router, this should not work?

Reply:
Previously, We only return the `zkAddress` and `kafkaBootstrapServers` when `d2.service.discovery.response.v2.enabled` is set to true in request headers. This is defined in D2ServiceDiscoveryResponseV2.java.

This was added when jackson in clients couldn't handle missing fields. We have made the code handle missing fields in json by default since https://github.com/linkedin/venice/commit/e8e13025.

There are 2 evidences to support that this won't break actual users:
1. We didn't face issues when we added `errorType`
2. We have added `serverD2Service` as a field in `D2ServiceDiscoveryResponse`

When we were implementing `serverD2Service`, I had verified that no one was using the legacy clients that couldn't handle unknown fields

Reply:
With this change, serializing and deserializing using `D2ServiceDiscoveryResponse` and `D2ServiceDiscoveryResponseV2` will work the same since `D2ServiceDiscoveryResponseV2 extends D2ServiceDiscoveryResponse` and doesn't add any fields of it's own